pkg/reconciler/testing/reconcile.go

```diff
@@ -0,0 +1,392 @@
+/*
+Copyright (c) 2020 TriggerMesh Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package testing
+
+import (
+	"testing"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	clientgotesting "k8s.io/client-go/testing"
+
+	"knative.dev/eventing/pkg/apis/eventing"
+	eventingv1beta1 "knative.dev/eventing/pkg/apis/eventing/v1beta1"
+	"knative.dev/pkg/apis"
+	duckv1 "knative.dev/pkg/apis/duck/v1"
+	rt "knative.dev/pkg/reconciler/testing"
+
+	"github.com/aws/aws-sdk-go/aws/arn"
+
+	"github.com/triggermesh/aws-event-sources/pkg/apis/sources/v1alpha1"
+	"github.com/triggermesh/aws-event-sources/pkg/reconciler/common"
+	eventtesting "github.com/triggermesh/aws-event-sources/pkg/reconciler/common/event/testing"
+)
+
+const (
+	tNs   = "testns"
+	tName = "test"
+	tKey  = tNs + "/" + tName
+	tUID  = types.UID("00000000-0000-0000-0000-000000000000")
+)
+
+var tSinkURI = &apis.URL{
+	Scheme: "http",
+	Host:   "default.default.svc.example.com",
+	Path:   "/",
+}
+
+type AdapterDeploymentBuilderFuncForSource func(v1alpha1.AWSEventSource) common.AdapterDeploymentBuilderFunc
+
+// Test the Reconcile method of the controller.Reconciler implemented by controllers.
+//
+// The environment for each test case is set up as follows:
+//  1. MakeFactory initializes fake clients with the objects declared in the test case
+//  2. MakeFactory injects those clients into a context along with fake event recorders, etc.
+//  3. A Reconciler is constructed via a Ctor function using the values injected above
+//  4. The Reconciler returned by MakeFactory is used to run the test case
+func TestReconcile(t *testing.T, ctor Ctor, src v1alpha1.AWSEventSource,
+	eventTypes []string, adapterFn AdapterDeploymentBuilderFuncForSource) {
+
+	arn, err := arn.Parse(src.GetARN())
+	if err != nil {
+		t.Fatalf("provided source object has invalid ARN: %s", err)
+	}
+
+	newEventSource := eventSourceCtor(src, arn, eventTypes)
+	newAdapterDeployment := adapterDeploymentCtor(adapterFn(newEventSource()), arn)
```

Comment:
I'm not proud of this mess.

The whole complexity of this series of function calls comes from the fact that we need to "prepare" the given source with a call to `eventSourceCtor()` before we can use it to generate an adapter Deployment. The reason for that is that `eventSourceCtor()` populates the source's metadata (namespace, name, uid) with constant attributes, and that metadata plays in role in the generation of the adapter Deployment (name, owner reference, ...).

Maybe it would be better to populate these _before_ passing the source to the test function. That would at least eliminate the call to `newEventSource()`.

Reply:
will it be helpful if we move `adapterDeploymentBuilder` from each source to common package and expose methods to set source-specific specs?

Reply:
Maybe, although it's not very "common" anymore in this case. I'll see how that could look like.

Reply:
I simplified those functions.

1. All sources must now be sent pre-populated to `TestReconcile`, so the call to `newEventSource` was removed.
2. `adapterDeploymentCtor` now receives `adapterFn, src`, which allows us to call `adapterFn(src)` _and_ read the ARN from `src`.